### PR TITLE
AUT-3602: Emit reauth_failed audit event in VerifyCodeHandler

### PR DESCRIPTION
**Pull request by alhcomer**
## What

- Emits the AUTH_REAUTH_FAILED audit event from the VerifyCodeHandler when a user has exceeded the max number of authentication attempts for any count type.
- Also verified in tests that there are no interactions with the authentication attempts service when it is not turned on.

## How to review

1. Code Review

### CODE REVIEW

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java [hunk 1]
```diff
@@ -263,7 +270,17 @@ private boolean checkErrorCountsForReauth(
                     ReauthAuthenticationAttemptsHelper.countTypesWhereUserIsBlockedForReauth(
                             countsByJourney, configurationService);
 
+            ClientRegistry client = userContext.getClient().orElse(null);
```
Comment (andrew-moores) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java:275]: What advantage does using orElse give if the client in the userContext is null?

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java [hunk 2]
```diff
@@ -263,7 +270,17 @@ private boolean checkErrorCountsForReauth(
...
             if (!countTypesWhereBlocked.isEmpty()) {
+                String rpPairwiseId =
+                        getInternalCommonSubjectIdentifier(userProfileMaybe.get(), client);
```
Comment (andrew-moores) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java]: If this is called with client having a value of null then you will get a NullPointerException.  If we know client is null then there is no point making this call as we are in an error state.

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java [hunk 3]
```diff
@@ -472,4 +489,15 @@ private JourneyType getJourneyType(
         }
         return journeyType;
     }
+
+    private String getInternalCommonSubjectIdentifier(
+            UserProfile userProfile, ClientRegistry client) {
+        var internalCommonSubjectIdentifier =
+                ClientSubjectHelper.getSubject(
```
Comment (andrew-moores) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java]: As mentioned above there is potential for escaped exceptions here.

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java [hunk 4]
```diff
@@ -217,7 +218,7 @@ public APIGatewayProxyResponseEvent handleRequestWithUserContext(
                     session,
                     codeRequest,
                     userContext,
-                    userProfileMaybe.isPresent() ? userProfileMaybe.get().getSubjectID() : null,
+                    userProfile != null ? userProfile.getSubjectID() : null,
```
Comment (andrew-moores) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java:221]: If the journey is reauth can we get to this point with a null subjectId?  Should we add this check to the guard at 161?

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java [hunk 5]
```diff
@@ -197,7 +198,7 @@ public APIGatewayProxyResponseEvent handleRequestWithUserContext(
                         codeRequest,
                         journeyType,
                         notificationType,
-                        userProfileMaybe,
+                        userProfile,
```
Comment (andrew-moores) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java:201]: We only need the subjectId in the `handleInvalidVerificationCode` method.

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java [hunk 6]
```diff
@@ -330,7 +330,7 @@ private void processSuccessfulCodeRequest(
         int loginFailureCount =
                 codeStorageService.getIncorrectMfaCodeAttemptsCount(session.getEmailAddress());
         var clientSession = userContext.getClientSession();
-        var clientId = userContext.getClient().get().getClientID();
+        var clientId = userContext.getClient().map(ClientRegistry::getClientID).orElse(null);
```
Comment (andrew-moores) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java:333]: What advantage does this give over simply calling getClient when the client is null?  Similar comments as previously about proceeding when the client is null.

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java (hunk 1 above)
Comment (alhcomer) [reply frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/VerifyCodeHandler.java:275]: Rather than returning an optional we get null

**Review by andrew-moores — DISMISSED** [review]

**Review by BeckaL — APPROVED** [review]

**Review by BeckaL — APPROVED** [review]

**Pull request merged**